File: 1048-longest-string-chain/1048-longest-string-chain.java
class Solution {
    public int longestStrChain(String[] words) {
        Map<String,Integer> map = new HashMap<>();
        Arrays.sort(words, (a,b)->a.length() - b.length());
        int res = 0;
        for(String word : words)
        {
            int best = 0;
            for(int i=0;i<word.length();i++)
            {
                String prev = word.substring(0,i) + word.substring(i+1);
                best = Math.max(best,map.getOrDefault(prev,0)+1);
            }
            map.put(word,best);
            res = Math.max(res,best);
        }
        return res;
    }
}

File: 1048-longest-string-chain/1048-longest-string-chain.cpp
bool cmp(string &a, string &b){
    return a.length() < b.length();
}
class Solution {
public:
    int longestStrChain(vector<string>& words) {
        sort(words.begin(), words.end(), cmp);
        map<string, int> mp;
        int ans = 0;
        for(auto word: words){
            int cur = 0;
            for(int i=0;i<word.length(); i++){
                string prev = word.substr(0,i)+word.substr(i+1);
                cur = max(cur, mp[prev]+1);
            }
            mp[word] = cur;
            ans = max(ans, cur);
        }

        return ans;
    }
};